design/design-doc-agent.md

```diff
@@ -0,0 +1,586 @@
+containers:
+- name: wordpress
+    image: "wordpress:4.6.1-apache"
+    env:
+    - name: WORDPRESS_DB_PASSWORD
+    valueFrom:
+        secretKeyRef:
+          name: sql-creds
+          key: password
+```
+
+The agent will be a Kubernetes controller running in the remote cluster and
+watching all `*Requirement` types. Next sections will talk about the
+implementation and user experience we'd like to have.
+
+### Synchronization
+
+In local mode, users directly interact with what's published by the platform
+team, which is `*Requirement` types and consume the infrastructure by mounting
+the secret whose name they specified on the `*Requirement` custom resource. To
+keep this experience, we need to have a synchronization loop for the following
+resources:
+
+* Pull
+  * `CustomResourceDefinition`s of all types that we want the applications to be
+    able to manage and view:
+    * Requirements that are published via `InfrastructurePublication`s. The
+      source of truth will be the remote cluster.
+    * `InfrastructureDefinition`, `InfrastructurePublication` and `Composition`.
+  * `Composition`s to discover what's available to choose. CRs of this type will
+    be read-only, and the source of truth will be the central cluster.
+  * `InfrastructurePublication`s to discover what's availabe as published API.
+    Read-only.
+  * `InfrastructureDefinition`s to discover how the secret keys are shaped.
+    Read-only.
+  * `Secret`s that are result of the infrastructure that is provisioned so that
+    it can be mounted to `Pod`s. Read-only.
+* Push
+  * `*Requirement` custom resources so that infrastructure can be requested.
+    Read and write permissions in a specific namespace in the central cluster will
+    be needed.
+
+### RBAC
+
+As we have two different Kubernetes cluster, there will be two separate security
+domains and because of that, the `ServiceAccount`s in the remote cluster will
+not be able to do any operation in the central cluster. Since the entity that
+will execute the operations in the central cluster is the agent, we need to
+define how we can deliver the necessary credentials to the agent so that it can
+connect to the central cluster. Additionally, it will need some permissions to
+execute operations in the remote cluster like `Secret` and
+`CustomResourceDefinition` creation. We will look at how the agent will be
+granted permissions to do its job in two separate domains with different
+mechanisms.
+
+#### Authenticating to The Central Cluster
+
+In order to execute any operation, a `ServiceAccount` needs to exist in the
+central cluster with appropriate permissions to read during pull and write
+during push operations while synchronizing with central cluster. Since the agent
+is running in the remote cluster, the credentials of this `ServiceAccount` will
+be stored in a `Secret` in the remote cluster. Alongside the credentials, the
+agent needs to know the namespace that it should sync to in the central cluster.
```

Comment:
You mean two different remote clusters with their own agents configured to target the same namespace in the central cluster, right? That's a good point. I see two options there:
* Annotate requirements in the central cluster with the UID of `kube-system` namespace of the remote cluster that requested it.
* Allow conflict to happen. There could be cases where several clusters use the same cloud services provisioned in the central cluster.
  * Maybe we can have a read-only mode for the agent? It would only pull from the central cluster instead of actively pushing.


I am kind of leaning towards the second option partly because same cloud service (like network or message queue) being used by several remote cluster sounds like a valid use case and partly because preventing this would be making a decision to restrict the user's ability to have different scenarios that we don't anticipate, i.e. the benefit of preventing it may not outweigh its downside, which is accidentally messing up the requirements with different remote clusters.

Reply:
Yeah, I think you call out below that only one agent is running per cluster which my initial question. i.e. it wouldn't make any sense to run one agent in foo, and another in baz namespaces because they both operate on the cluster scope and seems like a headache to manage.